examples/fixed_axis_tsne.py

```diff
@@ -0,0 +1,346 @@
+    dist = np.add(np.add(-2 * np.dot(x, x.T), sum_x).T, sum_x)
+    return dist
+
+
+def cal_perplexity(dist, idx=0, beta=1.0):
+    '''Calculate perplexity. D is distance vector
+    idx is the distance between a point and itself,
+    beta is Gaussian distribution parameter
+    '''
+    prob = np.exp(-dist * beta)
+
+    prob[idx] = 0
+    sum_prob = np.sum(prob)
+    if sum_prob < 1e-12:
+        prob = np.maximum(prob, 1e-12)
+        perp = -12
+    else:
+        perp = np.log(sum_prob) + beta * np.sum(dist * prob) / sum_prob
+        prob /= sum_prob
+
+    return perp, prob
+
+
+def seach_prob(x, tol=1e-5, perplexity=30.0):
+    '''Using binary research to find beta,
+    then calculate the prob of the pairwise
+    '''
+
+    # initialize parameters
+    print("Computing pairwise distances...")
+    (n, d) = x.shape
+    dist = cal_pairwise_dist(x)
+    dist[dist < 0] = 0
+    pair_prob = np.zeros((n, n))
+    beta = np.ones((n, 1))
+    # Here use the log value to make the later calculation easier
+    base_perp = np.log(perplexity)
+
+    for i in range(n):
+        if i % 500 == 0:
+            print("Computing pair_prob for point %s of %s ..." % (i, n))
+
+        betamin = -np.inf
+        betamax = np.inf
+        perp, this_prob = cal_perplexity(dist[i], i, beta[i])
+
+        # Using binary research to find the prob under the best sigma
+        perp_diff = perp - base_perp
+        tries = 0
+        while np.abs(perp_diff) > tol and tries < 50:
+            if perp_diff > 0:
+                betamin = beta[i].copy()
+                if betamax == np.inf or betamax == -np.inf:
+                    beta[i] = beta[i] * 2
+                else:
+                    beta[i] = (beta[i] + betamax) / 2
+            else:
+                betamax = beta[i].copy()
+                if betamin == np.inf or betamin == -np.inf:
+                    beta[i] = beta[i] / 2
+                else:
+                    beta[i] = (beta[i] + betamin) / 2
+
+            # update the value for perb and prob
+            perp, this_prob = cal_perplexity(dist[i], i, beta[i])
+            perp_diff = perp - base_perp
+            tries = tries + 1
+        # record the value for prob
+        pair_prob[i,] = this_prob
+    print("Mean value of sigma: ", np.mean(np.sqrt(1 / beta)))
+
+    return pair_prob
+
+
+def create_triangular_distance_matrix(
+        data):
+    distance_matrix = np.zeros(int((data.shape[0] + 1) * data.shape[0] / 2))
+    size = len(data)
+
+    k = 0
+    for i in range(size):
+        for j in range(i, size):
+            distance_matrix[k] = euclidean(data[i], data[j])
+            k = k + 1
+
+    return distance_matrix
+
+
+def euclidean(x, y):
+    """Standard euclidean distance.
+
+    ..math::
+        D(x, y) = \sqrt{\sum_i (x_i - y_i)^2}
+    """
+    result = 0.0
+    for i in range(x.shape[0]):
+        result += (x[i] - y[i]) ** 2
+    return np.sqrt(result)
+
+
+def kruskal_stress(distance_matrix, projection):
```

Comment:
could we use the kruskal stress already implemented in the cpu_file? If you copy-pasted this from somewhere, you'll need to also include the licence from the source

Reply:
I am copy & past it from our force scheme file. I am doing this because this is just an example and I tried to make it able to run individually and without any dependency. We can definitely change it to depend on the older file, if that's preferred. Also, the original tsne code is grabbed from internet and the source code is the github link I sent to you earlier in teams. I added the citation of the original code in the file.

Reply:
Yeah, it is better to use things already implemented. If we ever need to move or update it, then we would need to find all the different implementations and update them individually.

Reply:
license and reference added